File: 2800-minimum-string-length-after-removing-substrings/2800-minimum-string-length-after-removing-substrings.go
package main

import (
	"fmt"
)

func minLength(s string) int {
	// Initialize a stack as a slice of runes
	stack := []rune{}

	// Traverse the string `s`
	for _, c := range s {
		// Check if the stack is not empty and matches the pattern "AB" or "CD"
		if len(stack) > 0 {
			top := stack[len(stack)-1]
			if top == 'A' && c == 'B' {
				// Pop 'A' when "AB" is found
				stack = stack[:len(stack)-1]
				continue
			} else if top == 'C' && c == 'D' {
				// Pop 'C' when "CD" is found
				stack = stack[:len(stack)-1]
				continue
			}
		}
		// Otherwise, push the current character onto the stack
		stack = append(stack, c)
	}

	// Return the size of the remaining stack
	return len(stack)
}

File: 2800-minimum-string-length-after-removing-substrings/2800-minimum-string-length-after-removing-substrings.cpp
class Solution {
public:
    int minLength(string s) {
        stack<char> stack;

        // Traverse the input string
        for (int i = 0; i < s.length(); i++) {
            char cur_char = s[i];

            // If stack is empty, just push the current character
            if (stack.empty()) {
                stack.push(cur_char);
                continue;
            }

            // Check if the top of the stack forms "AB" or "CD" with cur_char
            if (cur_char == 'B' && stack.top() == 'A') {
                stack.pop(); // Remove the "AB" pair
            }
            else if (cur_char == 'D' && stack.top() == 'C') {
                stack.pop(); // Remove the "CD" pair
            }
            else {
                stack.push(cur_char); // Push the current character if no pair is formed
            }
        }

        // Return the remaining size of the stack
        return stack.size();
    }
};
